SPECIFICATIONS.md

```diff
@@ -424,6 +448,26 @@ Op | stack
 
 The stack contains only one value, and it is `true`: the expression succeeds.
 
+### Datalog fact generation
+
+Datalog fact generation works by repeatedly extending a Datalog _world_ until no new facts are generated.
+
+A Datalog world is:
+
+- a set of _rules_, each one tagged by the block id they were defined in
+- a set of _facts_, each one tagged by its _origin_: the block ids that allowed them to exist
```

Comment:
Note that facts can be represented grouped by origin, until they're filtered based on origin and matched individually. That's what the Haskell impl does, to reduce memory footprint and improve perf